### PR TITLE
Benchmark Multiraft.SubmitCommand.

### PR DESCRIPTION
**Pull request by bdarnell**
```
make bench PKG=./multiraft TESTFLAGS='-benchmem' 2>/dev/null
go test -tags ''  -i ./multiraft
go test -tags ''  -run - -cpu 1 -bench . ./multiraft -timeout 5m -benchmem
PASS
BenchmarkSubmitCommand1Node1Pending       100000         19396 ns/op        2519 B/op         30 allocs/op
BenchmarkSubmitCommand3Nodes1Pending        5000        326288 ns/op       20528 B/op        250 allocs/op
BenchmarkSubmitCommand5Nodes1Pending        3000        591363 ns/op       37505 B/op        453 allocs/op
BenchmarkSubmitCommand1Node2Pending       100000         13240 ns/op        1791 B/op         23 allocs/op
BenchmarkSubmitCommand3Nodes2Pending        5000        280565 ns/op       18846 B/op        229 allocs/op
BenchmarkSubmitCommand5Nodes2Pending        3000        524631 ns/op       35353 B/op        425 allocs/op
BenchmarkSubmitCommand1Node4Pending       100000         14925 ns/op        1989 B/op         24 allocs/op
BenchmarkSubmitCommand3Nodes4Pending        5000        260716 ns/op       18823 B/op        224 allocs/op
BenchmarkSubmitCommand5Nodes4Pending        3000        497292 ns/op       35304 B/op        418 allocs/op
ok      github.com/cockroachdb/cockroach/multiraft  14.784s
```

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3128)

### CODE REVIEW

File: multiraft/multiraft_test.go [hunk 1]
```diff
@@ -635,3 +635,90 @@ func TestConfigValidation(t *testing.T) {
...
+	groupID := roachpb.RangeID(1)
+	cluster.createGroup(groupID, 0, numNodes)
+	if numNodes > 1 {
+		cluster.elect(0, groupID)
+	}
+
+	// Rate limiting: the producer (main) thread writes to the throttle
+	// channel before submitting a command. The consumer thread started
+	// here reads from the throttle channel as commands are committed.
+	throttle := make(chan struct{}, pendingCommands)
```
Comment (tschottdorf) [inline multiraft/multiraft_test.go]: somewhat frail, but as long as you have no reasonable suspicion that unexpected messages could show up in the committed feed (due to reelections, etc), that's fine with me.

**Comment by tschottdorf** [comment]
LGTM. Any observations about the data? Do you think the slowdown from three to five nodes is general system load or are we serializing something when we shouldn't?

**Comment by bdarnell** [comment]
The slowdown from 3 to 5 nodes is in line with my expectations: for a 3-node cluster the leader sends messages to 2 followers; in a 5-node cluster there are 4 followers so twice as many messages.

A cpu profile shows we're spending about half our time in system calls in the transport layer. Not much we can do about that except sending fewer messages, but I don't think there are any wasted messages here (I should verify this, of course). It looks like there are some easy opportunities to reduce allocations.

File: multiraft/multiraft_test.go (hunk 1 above)
Comment (bdarnell) [inline multiraft/multiraft_test.go]: There shouldn't be any elections in this test since it is using a manual clock.

**Pull request merged**